### ISSUE TITLE
Import helpdb.jl and intergrate with REPL help.

### ISSUE THREAD

**Issue by MichaelHatherly** [MichaelHatherly/Lexicon.jl#6]
Load the contents of `helpdb.jl` and create `__METADATA__` objects in required modules. Also intercept REPL help system (namely `Base.Help.@help` macro) to call `Lexicon.query`.

**Comment by MichaelHatherly** [comment]
Already implemented. Not importing `helpdb.jl`, but rather commandeering the REPL's `?` mode.

**Comment by hayd** [comment]
Does this mean you ought to be able to apply Lexicon to Base?

e.g.
```
julia> save("base.html", Base)
ERROR: ArgumentError: Base is not documented.
 in metadata at /Users/andy/.julia/v0.4/Docile/src/interface.jl:57
```
*I get the feeling I'm missing something obvious here!*

**Comment by MichaelHatherly** [comment]
Nope, you're not missing anything. My closing comment was just a bit vague.

What's actually implemented is [this](https://github.com/MichaelHatherly/Lexicon.jl/blob/2d54864a2824c252d05c276638ca3fc8ab5c26e4/src/query.jl#L201-L204) and [this](https://github.com/MichaelHatherly/Lexicon.jl/blob/2d54864a2824c252d05c276638ca3fc8ab5c26e4/src/query.jl#L286-L298). Lexicon intercepts the `?` mode in the REPL, displays the base help docs related to a query and then displays any results from documented packages.

A while ago I did try parsing the base docs, but lack of an rst package to do the parsing of docstrings and signatures that didn't always match up to actual julia syntax meant it was hit or miss most of the time trying to link objects to docstrings.

**Comment by hayd** [comment]
Ah, ok thanks! Would be interesting to see if we can do something going forward (if Julia's migrating to md anyways). I wonder how lossy using pandoc to do the conversion would be... As a proof of concept, even if there's some/many conversion errors, it would be very cool! :)

**Comment by MichaelHatherly** [comment]
Not sure how pandoc would do with this. Would definitely be worth investigating.

**Comment by hayd** [comment]
You could use pypandoc and pycall, something like:

    pypandoc.convert(rst_string, 'rst', 'md', extra_args=('-tmarkdown_github',))

should have a good go at converting from rst to md (with the gh table syntax), I'm not sure what else will be missing...

**Comment by MichaelHatherly** [comment]
That's really cool! Does pypandoc include a pandoc binary with it or does that need to be downloaded separately?

**Comment by hayd** [comment]
I'm not sure tbh, I installed it seperately: [pandoc](http://johnmacfarlane.net/pandoc/installing.html) - make sure your pandoc version is 0.13+ so it has github flavour.

IIRC if pandoc is in your PATH it all "just works".

**Comment by hayd** [comment]
Another thing to support is [cross references](https://github.com/JuliaLang/julia/blob/48faa5a5df5329d48c3f287b0e8a1d60011f63df/test/markdown.jl#L34). cc @jiahao 

Should we make this a sep issue(s), I think there's kinda three here:

1. convert rst docs to Julia md without loss (tables + cross-references + testable code (?) + ...)\*, Note: we can apply plain to get it to something md/doc can parse.

2. get help from REPL (with current system)... I don't know if this is just me but if I `import Lexicon` help stops working in the REPL. `ERROR: UndefVarError: @help not defined` (e.g. after `?fft`.

3. get Lexicon working with Base (to generate docs).

---

Am I right in thinking once we have 1. and 2. we get 3. free? and 1. is desired in the long-term for Base.

 \*My thinking was that if we solve that we could write a script to paste this as `@doc` where the function is defined... though I don't see how that should quite work (due to ambiguity in where to put it)... although you could just dump all the docstrings in a single file as `@doc coverted_thing_found f` (hmmm that could actually work).

At the moment, I don't understand which (if any) of these is generated from the others: [help.db basedocs.jl math.rst](https://github.com/JuliaLang/julia/search?utf8=%E2%9C%93&q=%22simply+performs+this+operation+along+each+transformed+dimension%22&type=Code)

---

*I think we should be translating these as meta though @one-more-minute feels strongly the other way!*

**Comment by jiahao** [comment]
The primary source files are in `doc/stdlib/*.rst`. You can see for yourself what ReST/Sphinx features are used.

Running `make -C helpdb` updates `doc/helpdb.jl`.

`basedocs.jl` is new; I don't know how it is generated. @one-more-minute.

**Comment by jiahao** [comment]
My experience with `pandoc` is that `markdown -> html` is good, `latex -> html` is decent (if you have no equations), and other conversions I've tried (like `markdown -> markdown` to use `pandoc`'s citation generator) produced unsatisfactory results.

I once looked into trying to add in missing features [upstream](https://github.com/jgm/pandoc), but then I realized I had to learn Haskell first.

**Comment by one-more-minute** [comment]
`basedocs.jl` is the moral equivalent of `helpdb.jl`, but at the moment it's more a proof of concept than anything else – completely written out by hand.

Assuming we're agreed that Markdown docs in base should be the canonical format, the steps are:

* Generate `helpdb.jl`/`basedocs.jl`/whatever from the current RST docs, as a bunch of `@doc` invocations.

I don't see a way to make this lossless without basically writing an RST parser, so there's going to have to be some human intervention and/or horrible regexes involved. Luckily this only happens once. Also, it would be nice to move a lot of the docs inline, but since that's not possible in general something like `helpdb.jl` will always be necessary.

* Generate manual pages

Given that we're using RST at the moment, it might be right to just generate RST pages from Julia. We can do *that* losslessly, with references etc. We can also generate HTML directly, output Markdown for MkDocs, whatever – the world's our oyster, basically.

And that's it, really.

**Comment by one-more-minute** [comment]
Also, I should mention that I'm not at all against metadata, I just think it should be done via the Markdown itself. For example:
```julia
@doc meta(doc"Foo function", Dict(:author => "Mickey")) ->
function foo() ...
# or, pending syntax:
"""
Foo function
* Author: Mickey
"""
function foo() ...
```
I think it's pretty clear what's more convenient for the coder writing this, the maintainer reading it, and (most importantly) the helpless user looking at it in the REPL. The main concern here is being able to use the information programmatically, *but* that turns out to be easy thanks to the fact that we can easily parse markdown and pull metadata out. It's not the approach you'd take if you were looking for water-tightness against ambiguity, but the priority here has to be the human consumer (and, fortunately, humans are pretty good with syntactic ambiguity).

**Comment by hayd** [comment]
Happy to have a go here, is there an easy way to extract a Dict of methods to rsts?

I think we should put together the interesting cases in the rsts (e.g. cross-links, tables, etc), and throw them at pandoc and see what we lose:

1. file bug for X to pandoc and see if it's an easy fix
2. Count how many examples of X in the rsts there are + go through by hand/regex fix each X in the rst i.e. regex to find/fix.

----

Hmmmm interesting idea about munging out the meta by extending markdown. What about the following at the end of a docstring (although we don't support vbar yet):

```
Foo function
---
* author: Mickey
* return_type: Dict
```
translates/lowers to something with meta.
```
@doc meta(doc"Foo function", :author => "Mickey", :return_type => "Dict") ->
function foo() ...
```

**Comment by MichaelHatherly** [comment]
@one-more-minute, I've also been leaning towards embedding the metadata in the docstring itself recently myself. `meta` was a nice idea, but once I started writing out some actual docstrings with it it gets annoying pretty quickly. Was definitely worth experimenting with though.

@hayd, the syntax you've got there is basically YAML I think. I'd been pondering importing the YAML.jl package for that. It seems a little more reasonable than having to write out all the additional `meta` syntax explicitly. 

The `ERROR: UndefVarError: @help not defined` is from differences in what `?` calls in 0.3 and 0.4. 0.4 uses `@doc` now. So I'll push a fix for that later today.

**Comment by hayd** [comment]
@MichaelHatherly the problem with YAML is it's super sensitive (I'm not sure if that's the case with YAML.jl, but IIRC it's the spec that's sensitive) which IMO goes against markdown - which is a very forgiving language.

I guess the main driver for this is "what is allowed in the meta" ? If it's a single line, like the above, then it's easy, if it's multiline parsing (e.g. sections) then things gets complicated... (If they're not complicated then can we just use a special MD List.)

**Comment by MichaelHatherly** [comment]
That's true. Might be worth not tying ourselves to a specific syntax yet, and rather see what people begin to use naturally for denoting "metadata" in their docstrings.

**Comment by one-more-minute** [comment]
+:100:

**Comment by hayd** [comment]
I have a small script which does the extraction, rst -> md conversion, and creates a julia file with lots of `@doc`s for each of the rst stdlib files (not got it working on all of them yet, but don't think it'll be too bad).

One barrier I've got to which would be great to solve is how to `@doc` against the signature (I commented on this somewhere else, not sure where). At the moment I have:
```
@doc """Constuct a Date type by `Period` type parts. Arguments may be in any order. Date parts not provided will default to the value of `Dates.default(period)`.
""" Date(period::Period...) -> Date
```
but I want something like the following to parse:

```
@doc """Constuct a Date type by `Period` type parts. Arguments may be in any order. Date parts not provided will default to the value of `Dates.default(period)`.
---
* return_type: Date
""" Date(period::Period...)
```

---

I'm **-**:100: as (to flog this horse some more) IMO if we don't agree on something which will parse, everyone will be doing something random and it will be impossible (or, at least, a complete mess) to parse later....especially if we going to do something like the above in Base (and stuff like https://github.com/JuliaLang/julia/pull/9838).

**Comment by one-more-minute** [comment]
For Base docs, let's just put the signature as a first line:
```julia
@doc """
    Date(period::Period...) -> Date
Constuct a Date type by `Period` type parts. Arguments may be in any order.
Date parts not provided will default to the value of `Dates.default(period)`.
""" Date
```
That seems a lot cleaner to me than having a bullet at the end, and it'll look nicer for the end user (which has to be the priority at the moment). It's also very easy to pull the signature or return type out if necessary.

The main issue I have with setting a standard is that it means predicting everyone's use cases, and there's no way I'd feel confident doing that until use is a bit more widespread. If it's too cumbersome, you discourage people from writing docs, and if it's too lightweight, you restrict those who need more. I'd happily throw parseability out the window before taking either of those options.

Moreover, remember that right now we have a very smart and dedicated community of beta testers using this thing. I have no doubt that some interesting ideas will emerge that we'd never have thought of, and the last thing I'd want to do is suppress that.

**Comment by one-more-minute** [comment]
In fact, I hope @MichaelHatherly won't mind me quoting him to drive my point home:

> `meta` was a nice idea, but once I started writing out some actual docstrings with it it gets annoying pretty quickly.

Not long ago a lot of people were arguing strongly for `meta` – if we'd standardised earlier we could easily have included it. It's absolutely a bad idea to standardise any similarly nice ideas that haven't been widely tested in practice (and at this point, there aren't any that have).

**Comment by MichaelHatherly** [comment]
Writing the "human-readable" signature as the first line is a good idea. When a docstring is quite long having the signature at the top can help with context while reading. Short doctrings don't really need this though I've found.

In some cases I've come across Lexicon's automatic signature can be unhelpfully long. This doesn't happen in every case though. Having to type out a signature manually and then keep it in sync is something that we could probably avoid since we do store all the required information. It's *just* a matter of displaying a useful amount of information.

How about this for starters:

If the first element in a docstring's `.content` is a `Markdown.Code` then use that as the signature, otherwise automatically generate and display one based on the documented object.

We could probably add the return type automatically as well in most cases using `Base.return_types` unless the result is overly complex or meaningless.

It would then be up to the documenter to decide whether they would like the system to add a signature or create their own.

**Comment by hayd** [comment]
IMO there aren't many in practice __because__ there's no easy options yet, at least I don't see any. If there was something in the Docile docs explaining either:

1. how to parse additional syntax into meta (roll your own `@doc`)
2. some example meta syntaxes (use an existing one)

I agree it would be good if these could be tested in the wild before migrating to Base. But I strongly disagree that a lack of usage at this stage, without the above, indicates a lack of appetite.

---

It sucks not *attaching* to the signature (the return_types is not so important, and prepending it is no big deal, though I think if we we should ensure we can extract it later), you really want to attach it to a specific signature. Perhaps I'm confusing things by talking about a type constructor (Date) rather than a method?

To give a silly example, when you do:
```
?> *("a", "b")
```
you want the help for `+` with these types to come up (atm it doesn't) and this is only possible if the help messages are attached via signature. e.g. with some syntax like:

    @doc """"
    `Base.*(s, t)`
    Concatenate strings. The "*" operator is an alias to this
    function.

       julia> "Hello " * "world"
       "Hello world"
    """ *(s::AbstractString, t::AbstractString)

*Note: This is what happens if you `@doc` at definition time, e.g. with your own methods.*

**Comment by one-more-minute** [comment]
Actually, that's already implemented ;) Documentation is already attached to specific methods and it's easy in principle to grab the docstring for a specific method. It's not implemented for the repl help in Base yet but it's easy to do. That's why the main priority for the inline signature is to be human readable – the computable data already exists.

**Comment by one-more-minute** [comment]
> If the first element in a docstring's `.content` is a `Markdown.Code` then use that as the signature, otherwise automatically generate and display one based on the documented object.

This is roughly along the lines I've been thinking. It'd be easy to also have e.g.
```
    :sig
blah blah blah
```
as notation to let the doc system generate a signature for you.

**Comment by hayd** [comment]
@one-more-minute what's the syntax to attach it to a specific signature (after definition time)? When trying the above I get:

```
ERROR: UndefVarError: s not defined
```

(I know it works if you `@doc` at definition time, it's awesome!)

*Hmmm interesting idea to label sig (if you provide sig use that), perhaps if you instead provide return_type (the doc system generates, actual_sig + " -> " + return_type).*

**Comment by MichaelHatherly** [comment]
> what's the syntax to attach it to a specific signature (after definition time)?

I've not come across a need to attach docs to a `Method` after definition, but I guess you could extract the `Method` you need using:

```
julia> m = first(methods(+, (String, String)))
+(a::AbstractString,b::AbstractString) at none:1

julia> @doc "docs for +" m
+(a::AbstractString,b::AbstractString) at none:1

julia> @doc m
  docs for +
```

Not that elegant though.

With the plain docstrings syntax in Docile the following also works:

```
"docs for +"
(+, String, String)
```

**Comment by MichaelHatherly** [comment]
> @one-more-minute: This is roughly along the lines I've been thinking. It'd be easy to also have e.g.

> @hayd: if you instead provide return_type (the doc system generates, actual_sig + " -> " + return_type).

Those seems reasonable. Perhaps `[returns]` rather than `[return_type]` to shorten it slightly?

**Comment by hayd** [comment]
> I've not come across a need to attach docs to a Method after definition

The need (for me) is generating from the stdlib rsts the `@doc` statements. Currently I can get something like the above (which seems so close to something that could work!)...

@MichaelHatherly I had tried with the methods, but perhaps I am getting a little confused that:
```
?> a  #works

?> *("a", "b") # doesn't

julia> @doc *("a", "b")  # doesn't
```
but I think this is just something to do with Base lookups (i.e. it being a special case)?

---

That Docile syntax looks fantastic.

---

As a user I want to be able to do something like:

```
help?> *(x::Bool,y::Bool)
ERROR: UndefVarError: x not defined

help?> *(a, b, c, xs...)
ERROR: UndefVarError: b not defined
```
If this could work on the user side: use the above to lookup a specific method and `@doc` that. Then we could do the same on the documenting side (which I agree is an unusual case, usually you want to doc at definition time, but it'll make rsts->doc statements much easier/useful).

i.e. a macro to take the above expressions to the correct specific method ?

**Comment by hayd** [comment]
@MichaelHatherly I was trying something like the following (in `docs.jl`) to get the method if you pass it an expression like `:(f("s"::String,t::Int,u,:v,7,::Any))`:

```
function maybe_get_method(def)
    if :head in names(def) && def.head == :call
        sig = Any[]
        for arg in def.args[2:end]
            if isa(arg, Symbol)
                push!(sig, Any)
            elseif isa(arg, QuoteNode)
                push!(sig, typeof(arg.value))
            elseif isa(arg, Expr)
                assert(arg.head == :(::))
                push!(sig, eval(arg.args[end]))  # this needs to escape too
            else
                push!(sig, typeof(arg))
            end
        end
        sig = tuple(sig...) # getting sig works a treat (provided the types are in scope)

        f = quote @init $(esc(def.args[1])) end   #this is wrong
        possible_methods = methods(f, sig)
        length(possible_methods) == 0 && throw(ArgumentError("No method found with the given signature"))
        return possible_methods[1]  # potentially you want to apply doc to each method in this list??
    end
    def
end
```
This seems to get the signature (`(AbstractString,Int64,Any,Symbol,Int64,Any)`) correctly, but I'm not escaping correctly - tbh I have no idea what I'm doing with quote and `@init`... Is it possible to do this? (I suspect I'm missing something obvious!)

The issue is that f, and potentially the datatypes in the signature, is not defined in the local scope.

---

That way the syntax alluded to above ought to "just work" :)

**Comment by MichaelHatherly** [comment]
Sorry for the delayed response @hayd.

Where in `docs.jl` are you trying to call `maybe_get_method` from? The [`@init`](https://github.com/JuliaLang/julia/blob/100be8b517ff92ebe9dd9f91dfa90f6a4f049aa8/base/docs.jl#L13-L23) macro in

```julia
f = quote @init $(esc(def.args[1])) end   #this is wrong
```

needs to be spliced back into the location where `@doc` is called from so that `META` will refer to the correct `META`. I don't think it will be necessary to use it in this particular case.

`typeof(f) === Expr` and `methods` doesn't have a matching signature for that type. To get the actual `Function` object to pass to `methods` you could probably use something like:

```julia
f = getfield(current_module(), def.args[1])
possible_methods = methods(f, sig)
```

You might also be able to use `getfield(current_module(), ...)` to replace the use of `eval`.

**Comment by MichaelHatherly** [comment]
> The need (for me) is generating from the stdlib rsts the `@doc` statements.

Yes, that makes sense. Totally forgot what the title of this issue was.

**Comment by one-more-minute** [comment]
> The need (for me) is generating from the stdlib rsts the `@doc` statements.

For now it would be fine to just associate the whole docstring with the function itself, I think. At worst we're not losing anything over what's currently available. I'll also add a syntax for this to Base so that it can be extended in future.

**Comment by hayd** [comment]
@one-more-minute the annoying this is you have several docstrings (for the same method) then you have to just concat the strings (lose some info and gets a bit more complicated), if we can do the method signature thing this would be much nicer :)

**Comment by MichaelHatherly** [comment]
From a quick skim of the manual there doesn't seem to be too many `Method`-specific docs at the moment luckily, so @one-more-minute's suggestion should work for most of this. For the docs that are associated with particular methods it might be worth starting to write the docs above the actual methods. A syntax for what you'd like to accomplish @hayd would be nice to have though.

**Comment by hayd** [comment]
@MichaelHatherly I meant to say, your above worked a treat to get the method!

I need some hand holding, neither of the following quite work (in docm definition), it feels close...

    # setting
    isa(m, Method) && return doc(getfield(current_module(), m.func.code.name),
                                                    m, meta, m.func.code) 
    

    # retrieving
    isa(m, Method) && return doc(getfield(current_module(), m.func.code.name), m)

unless @one-more-minute's already putting something together? But thought I sat what I'd tried!

---

[dates](http://julia.readthedocs.org/en/latest/stdlib/dates/#dates-functions) was one of the examples I started with which has a few method specific docs.

**Comment by MichaelHatherly** [comment]
@hayd could you post a diff or link to the branch?

**Comment by hayd** [comment]
@MichaelHatherly this is very wip/messy (I left some printing in here too sorry) https://github.com/hayd/julia/commit/246314b2e5518db6c1960c7ba1b693649b9c51d3 

create docs for each file e.g. make_doc("collections")

**Comment by MichaelHatherly** [comment]
@hayd, I've written some in line notes in the branch you've linked. (Not sure if you'd get notified about them there.)
